modules/Blockchain.js

```diff
@@ -273,37 +402,163 @@ class Blockchain {
      * Calling this method will subscribe to Blockchain's event which will be
      * emitted globally using globalEmitter.
      * Callback function will be executed when the event is emitted.
-     * @param event Event to listen to
-     * @param callback function to be executed
+     * @param event - Name of event to listen to
+     * @param callback - Function to be executed
+     * @param blockchain_id - Blockchain to listen to
      * @returns {number | Object} Event handle
      */
-    async subscribeToEventPermanentWithCallback(event, callback) {
-        return this.blockchain.subscribeToEventPermanentWithCallback(event, callback);
+    async subscribeToEventPermanentWithCallback(event, callback, blockchain_id) {
+        const blockStartConditions = [];
+
+        if (blockchain_id) {
+            const implementation = this.blockchain.find(e => e.config.network_id === blockchain_id);
+
+            const startBlockNumber = await implementation.getCurrentBlock();
+            blockStartConditions.push({
+                blockchain_id,
+                block: { [Op.gte]: startBlockNumber },
+            });
+        } else {
+            const promises = [];
+
+            for (let i = 0; i < this.blockchain.length; i += 1) {
+                const implementation = this.blockchain[i];
+                promises.push(implementation.getCurrentBlock());
+            }
+            const currentBlocks = await Promise.all(promises);
+
+            for (let i = 0; i < currentBlocks.length; i += 1) {
+                const implementation = this.blockchain[i];
+
+                blockStartConditions.push({
+                    blockchain_id: implementation.config.network_id,
+                    block: { [Op.gte]: currentBlocks[i] },
+                });
+            }
+        }
+        const handle = setInterval(async () => {
+            const where = {
+                event,
+                finished: 0,
+                [Op.or]: blockStartConditions,
+            };
+
+            const eventData = await Models.events.findAll({ where });
+            if (eventData) {
+                eventData.forEach(async (data) => {
+                    try {
+                        callback({
+                            name: `eth-${data.event}`,
```

Comment:
This is blockchain specific code, is it possible to move it into specific implementation?